src/lua/protobuf_wireformat.lua

```diff
@@ -0,0 +1,147 @@
+local ffi = require('ffi')
+local int64_t = ffi.typeof('int64_t')
+local uint64_t = ffi.typeof('uint64_t')
+
+local WIRE_TYPE_VARINT = 0
+local WIRE_TYPE_I64 = 1
+local WIRE_TYPE_LEN = 2
+-- SGROUP (3) and EGROUP (4) are deprecated in proto3.
+local WIRE_TYPE_I32 = 5
+
+-- {{{ Helpers
+
+-- 32-bit IEEE 754 representation of the given number.
+local function as_float(value)
+    local p = ffi.new('float[1]')
+    p[0] = value
+    return ffi.string(ffi.cast('char *', p), 4)
+end
+
+-- 64-bit IEEE 754 representation of the given number.
+local function as_double(value)
+    local p = ffi.new('double[1]')
+    p[0] = value
+    return ffi.string(ffi.cast('char *', p), 8)
+end
+
+-- Encode an integral value as VARINT without a tag.
+-- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>.
+-- This is a helper function to encode tag and data values.
+local function encode_varint(value)
```

Comment:
I'm a bit worrying that we have documentation comments in different styles within one patchset. I would suggest a style that I personally use.

> Short description.
> \<empty line>
> Details.
> \<empty line>
> More details.

A short description is a kind of title. It is separated from the rest of the description by an empty line.

A description may contain several paragraphs. Paragraphs are separated from each other by empty lines.

In this case:

```lua
-- Encode an integral value as VARINT without a tag.
--
-- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>.
--
-- This is a helper function to encode tag and data values.
local function encode_varint(value)
    <...>
end
```

In my opinion it is a rule of thumb to separate paragraphs in a prose text, because it improves readability. Without the blank lines a text looks like a large paragraph with lines carried in random places.

Reply:
Brought all the comments to the same format